rviz2 randomly crashes while running with nav2 stack

### Basic info
RViz2 from frequently crashes during the use while running with Nav2 stack. The problem appears in random places even if all Nav2-related stuff was disabled (see minimal configuration file in reproduction section below). Crashes were [noted](https://github.com/ros-planning/navigation2/issues/2367) on Crystal and still appearing on latest Rolling.

### More info
Crashes appears irregularly in different places. All types of messages and backtraces were detected and sorted-out by their probability of occurrence:
1. Invalid Argument
```
terminate called after throwing an instance of 'std::system_error'
  what():  Invalid argument
```
2. Segfault
```
Thread 16 "rviz2" received signal SIGSEGV, Segmentation fault.
```
3. Assertion Robust
```
rviz2: ../nptl/pthread_mutex_lock.c:428: __pthread_mutex_lock_full: Assertion `e != ESRCH || !robust' failed.
--Type <RET> for more, q to quit, c to continue without paging--
```
4. Deadlock: RViz2 hangs during the use
5. Assertion Prio
```
[rviz2-4] rviz2: tpp.c:82: __pthread_tpp_change_priority: Assertion `new_prio == -1 || (new_prio >= fifo_min_prio && new_prio <= fifo_max_prio)' failed.
[ERROR] [rviz2-4]: process has died [pid 174985, exit code -6, cmd '/home/leha/ros2_rolling/install/rviz2/lib/rviz2/rviz2 -d /home/leha/nav2_ws/install/nav2_bringup/share/nav2_bringup/rviz/nav2_default_view.rviz --ros-args'].
```

For the types No.1-4 GDB backtraces were collected and attached in the archive: [backtraces.zip](https://github.com/ros2/rviz/files/6618073/backtraces.zip)
For type No.5 backtrace was not obtained (it was detected once). All of the problems above were detected by using the same scenario. Thus maybe (but not proven) all issue types have the same root cause.

### Steps to reproduce

Despite the problem is irregular and there is no 100% algorithm how to reproduce it, here are the steps allowing with more or less probability to get the problem:
1. Replace `navigation2/nav2_bringup/bringup/rviz/nav2_default_view.rviz` file with minimal repro-configuraiton attached to current issue: [nav2_default_view.rviz.zip](https://github.com/ros2/rviz/files/6618076/nav2_default_view.rviz.zip)
2. Run Nav2 simulation as usual:
```
ros2 launch nav2_bringup tb3_simulation_launch.py
```
3. In appeared RViz screen set robot pose by "2D Pose Estimate"
4. enable-disable many times the "Controller" checkbox as shown in the picture below:
![reproduction](https://user-images.githubusercontent.com/60094858/121212037-a9ce3080-c885-11eb-8c78-c78e5908cdc4.png)
5. If problem won't appear within ~30-40 seconds, close RViz and repeat starting from step No.2.

### Environment
* Ubuntu 20.04 x64
* ROS2 Distro: Rolling built from sources (`ros2.repos` from Jun 1st).
* Nav2 main and Nav2 dependencies were [built from sources](https://navigation.ros.org/build_instructions/index.html#manually-build-main). Nav2 commit: https://github.com/ros-planning/navigation2/commit/b63b11f8a638faf389d6b99911d95899abe1580b
* RViz, Qt, OGRE, OpenGl version as printed by rviz:
RViz commit: 94f6dd6958b2927a307893ff174f182dcc7afd0e
```
$ ros2 run rviz2 rviz2 -l
...
[INFO] [1623165832.507667729] [rviz2]: *-*-* OGRE Initialising
[INFO] [1623165832.507680536] [rviz2]: *-*-* Version 1.12.1 (Rhagorthua)
...
[INFO] [1623166434.886964038] [rviz2]: OpenGl version: 4.6 (GLSL 4.6)
```

Qt version is not printed by rviz:
```
$ qmake --version
Using Qt version 5.12.8 in /usr/lib/x86_64-linux-gnu
```
* If source build, which git commit?
The problem is long-standing detected since Crystal. Accused commit was not found.

--- a/tf2_ros/include/tf2_ros/async_buffer_interface.h
+++ b/tf2_ros/include/tf2_ros/async_buffer_interface.h
@@ -47,7 +47,7 @@ namespace tf2_ros
 
 class TransformStampedFuture : public std::shared_future<geometry_msgs::msg::TransformStamped>
 {
-  typedef std::shared_future<geometry_msgs::msg::TransformStamped> _Base_type;
+  using _Base_type = std::shared_future<geometry_msgs::msg::TransformStamped>;
 
 public:
   /// Constructor

--- a/tf2_ros/include/tf2_ros/async_buffer_interface.h
+++ b/tf2_ros/include/tf2_ros/async_buffer_interface.h
@@ -33,8 +33,10 @@
 #include <functional>
 #include <future>
 #include <string>
+#include <utility>
 
 #include "tf2_ros/visibility_control.h"
+#include "tf2/buffer_core.h"
 #include "tf2/time.h"
 #include "tf2/transform_datatypes.h"
 
@@ -43,7 +45,39 @@
 namespace tf2_ros
 {
 
-using TransformStampedFuture = std::shared_future<geometry_msgs::msg::TransformStamped>;
+class TransformStampedFuture : public std::shared_future<geometry_msgs::msg::TransformStamped>
+{
+  typedef std::shared_future<geometry_msgs::msg::TransformStamped> _Base_type;
+
+public:
+  /// Constructor
+  explicit TransformStampedFuture(_Base_type && future) noexcept
+  : _Base_type(std::move(future)) {}
+
+  /// Copy constructor
+  TransformStampedFuture(const TransformStampedFuture & ts_future) noexcept
+  : _Base_type(ts_future),
+    handle_(ts_future.handle_) {}
+
+  /// Move constructor
+  TransformStampedFuture(TransformStampedFuture && ts_future) noexcept
+  : _Base_type(std::move(ts_future)),
+    handle_(std::move(ts_future.handle_)) {}
+
+  void setHandle(const tf2::TransformableRequestHandle handle)
+  {
+    handle_ = handle;
+  }
+
+  tf2::TransformableRequestHandle getHandle() const
+  {
+    return handle_;
+  }
+
+private:
+  tf2::TransformableRequestHandle handle_ {};
+};
+
 using TransformReadyCallback = std::function<void (const TransformStampedFuture &)>;
 
 /**
@@ -75,6 +109,13 @@ public:
     const tf2::TimePoint & time,
     const tf2::Duration & timeout,
     TransformReadyCallback callback) = 0;
+
+  /**
+   * \brief Cancel the future to make sure the callback of requested transform is clean.
+   * \param ts_future The future to the requested transform.
+   */
+  virtual void
+  cancel(const TransformStampedFuture & ts_future) = 0;
 };  // class AsyncBufferInterface
 
 }  // namespace tf2_ros

--- a/tf2_ros/include/tf2_ros/async_buffer_interface.h
+++ b/tf2_ros/include/tf2_ros/async_buffer_interface.h
@@ -47,21 +47,21 @@ namespace tf2_ros
 
 class TransformStampedFuture : public std::shared_future<geometry_msgs::msg::TransformStamped>
 {
-  using _Base_type = std::shared_future<geometry_msgs::msg::TransformStamped>;
+  using BaseType = std::shared_future<geometry_msgs::msg::TransformStamped>;
 
 public:
   /// Constructor
-  explicit TransformStampedFuture(_Base_type && future) noexcept
-  : _Base_type(std::move(future)) {}
+  explicit TransformStampedFuture(BaseType && future) noexcept
+  : BaseType(std::move(future)) {}
 
   /// Copy constructor
   TransformStampedFuture(const TransformStampedFuture & ts_future) noexcept
-  : _Base_type(ts_future),
+  : BaseType(ts_future),
     handle_(ts_future.handle_) {}
 
   /// Move constructor
   TransformStampedFuture(TransformStampedFuture && ts_future) noexcept
-  : _Base_type(std::move(ts_future)),
+  : BaseType(std::move(ts_future)),
     handle_(std::move(ts_future.handle_)) {}
 
   void setHandle(const tf2::TransformableRequestHandle handle)

--- a/tf2_ros/include/tf2_ros/buffer.h
+++ b/tf2_ros/include/tf2_ros/buffer.h
@@ -252,6 +252,14 @@ public:
       callback);
   }
 
+  /**
+   * \brief Cancel the future to make sure the callback of requested transform is clean.
+   * \sa cancel(TransformStampedFuture& ts_future);
+   */
+  TF2_ROS_PUBLIC
+  void
+  cancel(const TransformStampedFuture & ts_future) override;
+
   TF2_ROS_PUBLIC
   inline void
   setCreateTimerInterface(CreateTimerInterface::SharedPtr create_timer_interface)

--- a/tf2_ros/include/tf2_ros/buffer.h
+++ b/tf2_ros/include/tf2_ros/buffer.h
@@ -254,7 +254,7 @@ public:
 
   /**
    * \brief Cancel the future to make sure the callback of requested transform is clean.
-   * \sa cancel(TransformStampedFuture& ts_future);
+   * \param ts_future The future to the requested transform.
    */
   TF2_ROS_PUBLIC
   void

--- a/tf2_ros/include/tf2_ros/buffer_interface.h
+++ b/tf2_ros/include/tf2_ros/buffer_interface.h
@@ -49,8 +49,6 @@
 
 namespace tf2_ros
 {
-using TransformStampedFuture = std::shared_future<geometry_msgs::msg::TransformStamped>;
-using TransformReadyCallback = std::function<void (const TransformStampedFuture &)>;
 
 inline builtin_interfaces::msg::Time toMsg(const tf2::TimePoint & t)
 {

--- a/tf2_ros/include/tf2_ros/message_filter.h
+++ b/tf2_ros/include/tf2_ros/message_filter.h
@@ -439,9 +439,11 @@ public:
         buffer_timeout_,
         std::bind(&MessageFilter::transformReadyCallback, this, std::placeholders::_1, handle));
 
+      // If handle of future is 0 or 0xffffffffffffffffULL, waitForTransform have already called
+      // the callback.
       if (0 != future.getHandle() || 0xffffffffffffffffULL != future.getHandle()) {
         std::unique_lock<std::mutex> lock(ts_futures_mutex_);
-        ts_futures_.insert({handle, future});
+        ts_futures_.insert({handle, std::move(future)});
       }
     }
   }

--- a/tf2_ros/include/tf2_ros/message_filter.h
+++ b/tf2_ros/include/tf2_ros/message_filter.h
@@ -44,6 +44,7 @@
 #include <tuple>
 #include <type_traits>
 #include <unordered_map>
+#include <utility>
 #include <vector>
 
 #include "message_filters/connection.h"

--- a/tf2_ros/include/tf2_ros/message_filter.h
+++ b/tf2_ros/include/tf2_ros/message_filter.h
@@ -43,6 +43,7 @@
 #include <string>
 #include <tuple>
 #include <type_traits>
+#include <unordered_map>
 #include <vector>
 
 #include "message_filters/connection.h"
@@ -335,6 +336,14 @@ public:
    */
   void clear()
   {
+    {
+      std::unique_lock<std::mutex> lock(ts_futures_mutex_);
+      for (auto & kv : ts_futures_) {
+        buffer_.cancel(kv.second);
+      }
+      ts_futures_.clear();
+    }
+
     std::unique_lock<std::mutex> unique_lock(messages_mutex_);
 
     TF2_ROS_MESSAGEFILTER_DEBUG("%s", "Cleared");
@@ -423,12 +432,17 @@ public:
       const auto & handle = std::get<0>(param);
       const auto & stamp = std::get<1>(param);
       const auto & target_frame = std::get<2>(param);
-      buffer_.waitForTransform(
+      tf2_ros::TransformStampedFuture future = buffer_.waitForTransform(
         target_frame,
         frame_id,
         stamp,
         buffer_timeout_,
         std::bind(&MessageFilter::transformReadyCallback, this, std::placeholders::_1, handle));
+
+      if (0 != future.getHandle() || 0xffffffffffffffffULL != future.getHandle()) {
+        std::unique_lock<std::mutex> lock(ts_futures_mutex_);
+        ts_futures_.insert({handle, future});
+      }
     }
   }
 
@@ -484,6 +498,14 @@ private:
 
     MEvent saved_event;
     bool event_found = false;
+
+    {
+      std::unique_lock<std::mutex> lock(ts_futures_mutex_);
+      auto iter = ts_futures_.find(handle);
+      if (iter != ts_futures_.end()) {
+        ts_futures_.erase(iter);
+      }
+    }
 
     {
       // We will be accessing and mutating messages now, require unique lock
@@ -748,6 +770,13 @@ private:
 
   // Timeout duration when calling the buffer method 'waitForTransform'
   tf2::Duration buffer_timeout_;
+
+  ///< The mutex used for locking TransformStampedFuture map operations
+  std::mutex ts_futures_mutex_;
+
+  ///< Store the TransformStampedFuture returned by 'waitForTransform',
+  // to clear the callback in the Buffer if MessageFiltered object is destroyed.
+  std::unordered_map<uint64_t, tf2_ros::TransformStampedFuture> ts_futures_;
 };
 }  // namespace tf2_ros
 

--- a/tf2_ros/src/buffer.cpp
+++ b/tf2_ros/src/buffer.cpp
@@ -240,6 +240,7 @@ Buffer::waitForTransform(
     };
 
   auto handle = addTransformableRequest(cb, target_frame, source_frame, time);
+  future.setHandle(handle);
   if (0 == handle) {
     // Immediately transformable
     geometry_msgs::msg::TransformStamped msg_stamped = lookupTransform(
@@ -264,6 +265,18 @@ Buffer::waitForTransform(
     timer_to_request_map_[timer_handle] = handle;
   }
   return future;
+}
+
+void
+Buffer::cancel(const TransformStampedFuture & ts_future)
+{
+  cancelTransformableRequest(ts_future.getHandle());
+
+  std::lock_guard<std::mutex> lock(timer_to_request_map_mutex_);
+  auto iter = timer_to_request_map_.find(ts_future.getHandle());
+  if (iter != timer_to_request_map_.end()) {
+    timer_to_request_map_.erase(iter);
+  }
 }
 
 void

